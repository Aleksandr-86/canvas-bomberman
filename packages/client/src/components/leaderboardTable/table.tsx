import { useSortableData } from '../../hooks/userSortableData'

import styles from './table.module.css'

interface Props {
  playersStats: {
    id: number
    place: number
    name: string
    score: number
    games: number
    totalTime: number
  }[]
}

const headers = [
  ['place', 'Место'],
  ['name', 'Имя игрока'],
  ['score', 'Очки'],
  ['games', 'Игр'],
  ['totalTime', 'Время (мин)'],
]

export const Table = (props: Props) => {
  const { items, requestSort, sortConfig } = useSortableData(
    props.playersStats,
    {
      key: 'place',
      direction: 'ascending',
    }
  )

  const getClassNamesFor = (name: string) => {
    if (!sortConfig) {
      return
    }

    const { key, direction } = sortConfig

    if (key === name) {
      return `${styles.head} ${styles[direction]}`
    } else {
      return styles.head
    }
  }

  return (
    <table className={styles.table}>
      <thead>
        <tr>
          {headers.map(([headerName, headerTitle], index) => (
            <th
              key={index}
              onClick={() => requestSort(headerName)}
              className={getClassNamesFor(headerName)}>
              {headerTitle}
            </th>
          ))}
        </tr>
      </thead>

      <tbody>
        {items.map(item => (
          <tr className={styles.row} key={item.id}>
            <td className={styles.columnPlace}>{item.place}</td>
            <td className={styles.columnName}>{item.name}</td>
            <td className={styles.columnScore}>{item.score}</td>
            <td className={styles.columnGames}>{item.games}</td>
            <td className={styles.columnTotalTime}>{item.totalTime}</td>
          </tr>
        ))}
      </tbody>
    </table>
  )
}
