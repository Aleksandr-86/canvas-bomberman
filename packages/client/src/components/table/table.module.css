.table {
  width: 100%;
  margin-bottom: 30px;
}

.table__title {
  font-family: 'Rubik Bubbles', fantasy;
  font-weight: 400;
  font-size: 40px;
  line-height: 47px;
  text-align: center;
}
