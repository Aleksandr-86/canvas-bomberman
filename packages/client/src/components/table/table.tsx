import { useSortableData } from '../../hooks/userSortableData'

import styles from './table.module.css'

interface Props {
  caption: string
  userStats: {
    id: number
    place: number
    name: string
    score: number
    games: number
    totalTime: number
  }[]
}

const headers = [
  ['place', 'Место'],
  ['name', 'Имя игрока'],
  ['score', 'Очки'],
  ['games', 'Игр'],
  ['totalTime', 'Время (мин)'],
]

export const Table = (props: Props) => {
  const { items, requestSort, sortConfig } = useSortableData(props.userStats, {
    key: 'place',
    direction: 'ascending',
  })

  const getClassNamesFor = (name: string) => {
    if (!sortConfig) {
      return
    }

    const { key, direction } = sortConfig

    if (key === name) {
      return `${styles.table__head} ${styles[direction]}`
    } else {
      return styles.table__head
    }
  }

  return (
    <table className={styles.table}>
      <caption className={styles.table__caption}>{props.caption}</caption>

      <thead>
        <tr>
          {headers.map((header, index) => (
            <th
              key={index}
              onClick={() => requestSort(header[0])}
              className={getClassNamesFor(header[0])}>
              {header[1]}
            </th>
          ))}
        </tr>
      </thead>

      <tbody>
        {items.map(item => (
          <tr className={styles.table__row} key={item.id}>
            <td className={styles.table__column_place}>{item.place}</td>
            <td className={styles.table__column_name}>{item.name}</td>
            <td className={styles.table__column_score}>{item.score}</td>
            <td className={styles.table__column_games}>{item.games}</td>
            <td className={styles.table__column_totalTime}>{item.totalTime}</td>
          </tr>
        ))}
      </tbody>
    </table>
  )
}
