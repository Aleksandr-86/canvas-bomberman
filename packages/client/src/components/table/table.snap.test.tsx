import * as renderer from 'react-test-renderer'
import { Table } from './table'

const mockProps = {
  headers: ['Автор', 'Тема', 'Дата'],
  rows: [
    {
      id: 1,
      author: 'Kan88',
      body: 'Как пройти первый уровень',
      date: '8.01.22',
    },
    {
      id: 2,
      author: 'sasha85',
      body: 'Как создать новую тему',
      date: '8.01.22',
    },
  ],
}

describe(`<Table/> snapshot test cases`, () => {
  test(`<Table/> should render <table className=table> parent with child elements (with mockProps)`, () => {
    const tree = renderer.create(<Table {...mockProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
