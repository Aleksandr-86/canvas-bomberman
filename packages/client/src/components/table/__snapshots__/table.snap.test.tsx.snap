// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table/> snapshot test cases <Table/> should render <table className=table> parent with child elements (with mockProps) 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th
        className="tableTitle"
      >
        Автор
      </th>
      <th
        className="tableTitle"
      >
        Тема
      </th>
      <th
        className="tableTitle"
      >
        Дата
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        className="tableRow"
      >
        Kan88
      </td>
      <td
        className="tableRow"
      >
        Как пройти первый уровень
      </td>
      <td
        className="tableRow"
      >
        8.01.22
      </td>
    </tr>
    <tr>
      <td
        className="tableRow"
      >
        sasha85
      </td>
      <td
        className="tableRow"
      >
        Как создать новую тему
      </td>
      <td
        className="tableRow"
      >
        8.01.22
      </td>
    </tr>
  </tbody>
</table>
`;
